Feature/order

johnnyhelloworld opened the pull request:
# Pull Request Templates

## Type of change

- [X] Feature
- [ ] Fix

## Description

Order, OrderItems, Cart

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

Commit:
implement Order

Commit:
implement Cart

Commit:
implement OrderItems

Commit:
add subtotal in orderItems

Commit:
add priceEtx and priceVat in Order

Commit:
add shippingAddress and billingAddress in Order

DaniOmer commented on src/apps/cart/data-access/cart.interface.ts:
```diff
@@ -0,0 +1,10 @@
+
+export interface ICart extends Timestamps, Document {
+  _id: string;
+  user_id: Types.ObjectId;
```
Je pense que tu pourrais rajouter qu'il vaut mieux avoir `user: ObjectId | IUser` pour les permettre  le populate si besoin

DaniOmer commented on src/apps/cart/api/cart.controller.ts:
```diff
@@ -0,0 +1,38 @@
+import { ApiResponse } from "../../../librairies/controllers/api.response";
+
+export class CartController {
+  private cartService: CartService;
```
Il faudrait extend `BaseController` pour pouvoir utiliser `dataToDtoInstance`

DaniOmer commented on src/apps/cart/data-access/cartItem.interface.ts:
```diff
@@ -0,0 +1,7 @@
+import { Types } from "mongoose";
+
+export interface ICartItem {
+  orderItem_id: Types.ObjectId;
```
pareil utiliser IOrderItem ici pour permettre le populate si besoin

DaniOmer commented on src/apps/order/data-access/order.interface.ts:
```diff
@@ -0,0 +1,21 @@
+  status: OrderStatus;
+  priceEtx: number;
+  priceVat: number;
+  user_id: Types.ObjectId;
```
mieux vaut utiliser un champs user et non userId et rajouter IUser pour le populate

DaniOmer commented on src/apps/order/domain/order.service.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  async createOrder(data: Partial<IOrder>): Promise<IOrder> {
+    return this.repository.create(data);
```
Utilise CreateOrderDTO et non pas partiel<IOder> pour une validation strict

DaniOmer commented on src/apps/orderItems/domain/orderItems.service.ts:
```diff
@@ -0,0 +1,26 @@
+    const orderItem = await this.repository.getById(id);
+    if (!orderItem) {
+      throw new Error(`OrderItem with ID ${id} not found`);
+    }
```
pourquoi tu throw une Error et non une BadRequestError ? le middleware de gestion d'erreur gère les erreurs simples comme des erreur 500, à éviter du coup

DaniOmer commented on tsconfig.json:
```diff
@@ -15,7 +15,7 @@
     "experimentalDecorators": true /* Enable experimental support for legacy experimental decorators. */,
-    // "emitDecoratorMetadata": true,                    /* Emit design-type metadata for decorated declarations in source files. */
+    "emitDecoratorMetadata": true,                    /* Emit design-type metadata for decorated declarations in source files. */
     // "jsxFactory": "",                                 /* Specify the JSX factory function used when targeting React JSX emit, e.g. 'React.createElement' or 'h'. */
```
pourquoi tu as modifié le tsconfig ?

DaniOmer commented on src/apps/cart/domain/cart.service.ts:
```diff
@@ -0,0 +1,28 @@
+
+  async deleteCartByUserId(userId: string): Promise<ICart> {
+    const deletedCart = await CartModel.findOneAndDelete({ userId });
+    return this.validateDataExists(deletedCart, userId);
```
ici je penses qu'il faut throw une BadRequestError au lieu `this.validateDataExists`

DaniOmer commented:
sur la PR de @FAEZ10 j'ai fait la même remarque par rapport à `this.validateDataExists` qui est impertinent à mon avis. sur certains throw il serait intéressant de rajouter un context, ce qu'on ne peux faire avec `this.validateDataExists` (par exemple le middleware de validation d'id renvoie le même message d'erreur que les methods dans les services, Impossible de savoir en lisant les logs si l'erreur provient d'un update, d'une suppression, etc...) J'insiste depuis plusieurs jours parce que ça peut induire en erreur si besoin de debug ou même en lisant les logs par exemple mais en vrai si vous êtes tous okay pour utiliser ça par endroit je ne dirai plus rien sur les review à ce propos

DaniOmer commented:
Certains des tes services extends BaseController et d'autres non. ce serait bien de check ça, aussi si tu peux rajoute this.logger dans BaseService pour qu'il soit accessible dans tous les services, tu peux t'inspirer de BaseController

Commit:
refacto

DaniOmer commented on src/apps/orderItems/data-access/orderItems.repository.ts:
```diff
@@ -0,0 +1,21 @@
+import { IOrderItem } from "./orderItems.interface";
+
+export class OrderItemRepository {
+  async create(data: Partial<IOrderItem>): Promise<IOrderItem> {
```
Utiliser le DTO de création de OrderItem `(data : OrderItemCreateDTO)`

Commit:
regroup Cart, Order and OrderItems in Order repository

Commit:
add product_id + render product_id and productVariant_id nullable

DaniOmer commented on src/apps/order/cart/api/cart.route.ts:
```diff
@@ -0,0 +1,58 @@
+ *       200:
+ *         description: Cart retrieved successfully
+ */
+router.get("/", cartController.getCart.bind(cartController));
```
Il manque le middleware d'authentication

DaniOmer commented on src/apps/order/cart/api/cart.controller.ts:
```diff
@@ -0,0 +1,40 @@
+  async getCart(req: Request, res: Response, next: NextFunction): Promise<void> {
+    try {
+      const cart = await this.cartService.getCartByUserId(req.user.id);
+      ApiResponse.success(res, "Cart retrieved successfully", cart, 200);
```
je ne dois pas pouvoir récupérer la cart d'un autre. il faudrait une logique pour check que je suis bien l'auteur de la ressource demandé en gardant à l'esprit qu'un admin peut récupérer la cart de n'importe quel user

DaniOmer commented on src/apps/order/order/domain/order.service.ts:
```diff
@@ -0,0 +1,60 @@
+
+  async updateOrderById(
+    id: string,
+    updates: Partial<IOrder>
```
soit un omit  de IOrder ou soit `OrderUpdateDTO`

DaniOmer commented on src/apps/order/orderItems/api/orderItems.route.ts:
```diff
@@ -0,0 +1,84 @@
+
+const router = Router();
+const orderItemController = new OrderItemController();
+
```
Pourquoi tu as besoin d'un controller OrderItemController ?

DaniOmer commented on src/apps/order/cart/domain/cart.service.ts:
```diff
@@ -0,0 +1,50 @@
+
+  async updateCartByUserId(userId: string, updates: Partial<ICart>): Promise<ICart> {
+    const updatedCart = await CartModel.findOneAndUpdate({ userId }, updates, { new: true });
+    if (!updatedCart) {
```
tu update la cart, mais qu'en est-il des produits qui sont dans la cart ?

DaniOmer commented on src/apps/order/cart/domain/cart.service.ts:
```diff
@@ -0,0 +1,50 @@
+
+  async createCart(data: Partial<ICart>): Promise<ICart> {
+    return CartModel.create(data);
+  }
```
Pourquoi tu utilise CartModel et non pas le repository pattern ?
aussi tu créer la cart, mais qu'en est-il des items de la cart ?

DaniOmer commented on src/apps/order/order/domain/order.dto.ts:
```diff
@@ -0,0 +1,75 @@
+}
+
+export class OrderCreateDTO {
+  @IsNotEmpty()
```
il manque la liste des articles associés dans ton DTO

DaniOmer commented on src/apps/order/orderItems/domain/orderItems.dto.ts:
```diff
@@ -0,0 +1,43 @@
+import { Schema } from "mongoose";
+
+export class OrderItemCreateDTO {
+  @IsNotEmpty()
```
Pourquoi tu as besoin de ça ?

Commit:
add authentication middleware

Commit:
add logic user can only have access, update or delete his cart

Commit:
add logic user can only have access, update or delete his order

Commit:
add logic user can only have access, update or delete his orderItems

Commit:
omit updateOrderById

Commit:
add logic user can have access, update or delete his items in his cart

Commit:
add relation between Order and OrderItems

DaniOmer commented on src/apps/order/cart/api/cart.controller.ts:
```diff
@@ -0,0 +1,60 @@
+  async getCart(req: Request, res: Response, next: NextFunction): Promise<void> {
+    try {
+      const cart = await this.cartService.getCartByUserId(req.user.id);
+      if (!cart || cart.user.toString() !== req.user.id) {
```
ici l'admin doit aussi pouvoir récupérer la cart s'il le souhaite

DaniOmer commented on src/apps/order/cart/api/cart.controller.ts:
```diff
@@ -0,0 +1,60 @@
+  async updateCart(req: Request, res: Response, next: NextFunction): Promise<void> {
+    try {
+      const updatedCart = await this.cartService.updateCartByUserId(req.user.id, req.body);
+      if (!updatedCart || updatedCart.user.toString() !== req.user.id) {
```
pourquoi tu fais d'abord l'update avant de check l'utilisateur ? pareil que pour le get l'admin devrait avoir la possibilité de faire l'update

DaniOmer commented on src/apps/order/cart/api/cart.controller.ts:
```diff
@@ -0,0 +1,60 @@
+  async deleteCart(req: Request, res: Response, next: NextFunction): Promise<void> {
+    try {
+      const cart = await this.cartService.getCartByUserId(req.user.id);
+      if (!cart || cart.user.toString() !== req.user.id) {
```
same comment as above

DaniOmer committed:
Implemented some logic related to cart by introducing the following methods:
- getCartByUser to get current user cart
- addItemToCart to add item to cart
- updateCartItemQuantity to update existing cart item quantity
- clearCart to empty user cart

DaniOmer committed:
Added the logic to retrieve cart items on request to get cart

DaniOmer committed:
Implemented the workflow related to user address by introducing the following methods:
- createAddress to create address for billing or invoice
- updateAddress to update existing address
- getAllAddressesByTypeAndUserId to retrieve addresses by type (billing or invoice)

DaniOmer committed:
Implemented the workflow to create order from cart

DaniOmer committed:
Updated user cart workflow to introduce product reservation at cart level

DaniOmer committed:
Updated reserveStock method to handle reservation on cart item update

DaniOmer committed:
Updated releaseStock to fix product reservation overlap

DaniOmer committed:
Upated order workflow to require user access to protected resources

DaniOmer committed:
Updated cart service to fix issue related to cleartCart method not updating reserved stocks

DaniOmer committed:
Added the logic to create order items on order create

DaniOmer committed:
Implemented utility function to generate order number

DaniOmer committed:
Implemented logic to get order by status and by order number

DaniOmer committed:
Added changes related to CORS and Rate limit

DaniOmer committed:
Updated basic register to remove authType from DTO

DaniOmer committed:
Updated user register method to dynamically asign user role

DaniOmer committed:
Updated auth logic to generate email validation link to match frontend expectations
